### PR TITLE
POST /tasks fails when spending quota is reached

### PR DESCRIPTION
**Pull request by alongalky**

### CODE REVIEW

File: routes/api/tasks.js [hunk 1]
```diff
@@ -2,6 +2,49 @@ const moment = require('moment')
...
+    end = moment(task.timestamp_done)
+  } else {
+    end = moment()
+  }
+
+  task.durationInSeconds = end.diff(start, 'seconds')
+  return task
+}
+
+const addCost = tiers => task => {
```
Comment (karimodm) [inline routes/api/tasks.js]: what's the reason of double lambda? Do you plan to use only the first closure `addCost(tiers)` for something?
You only really use it on line 46 passing the task returned by `addTaskDuration`.
The form with two arguments `addCost(tiers, addTaskDuration(task))` would be more readable.

File: routes/api/tasks.js [hunk 2]
```diff
@@ -2,6 +2,49 @@ const moment = require('moment')
...
+
+    const err = new Error(`Spending quota exceeded`)
+    err.type = 'spending_quota_exceeded'
+    return Promise.reject(err)
+  } else {
+    return Promise.resolve()
+  }
+}
+
+const addTaskDuration = task => {
```
Comment (karimodm) [inline routes/api/tasks.js]: Having these functions that extend the object returned from the database with more calculated fields are nice. I think they belong to the database module though. I would make the `database.getTasks` method return task objects with this calculated data. This is especially true if other modules want to look into this data as well.

File: routes/api/tasks.js [hunk 3]
```diff
@@ -2,6 +2,49 @@ const moment = require('moment')
...
+  } else {
+    end = moment()
+  }
+
+  task.durationInSeconds = end.diff(start, 'seconds')
+  return task
+}
+
+const addCost = tiers => task => {
+  const tier = tiers.find(t => t.name === task.tier)
```
Comment (karimodm) [inline routes/api/tasks.js]: I am thinking that we will need to store config historical cost information about every tier we supported in Eureka for this number to keep consistent. Imagine we want to replace a tier spec for something else, these value will drift away. This comment is just a call for thoughts. Storing tier price per hour in the database on each task being inserted? This way configuration tiers can change but historical costs are preserved.

File: routes/api/tasks.js [hunk 4]
```diff
@@ -2,6 +2,49 @@ const moment = require('moment')
...
+  if (!tier) {
+    throw new Error(`Invalid tier found for task ${task.task_id}`)
+  }
+
+  task.costInCents = tier.pricePerHourInCents * (task.durationInSeconds / (60.0 * 60.0))
+  return task
+}
+
+const addTaskCostAndDuration = tiers => task => addCost(tiers)(addTaskDuration(task))
```
Comment (karimodm) [inline routes/api/tasks.js]: Double lambda makes sense here because of the .map, but not above. Elegant :+1:

File: test/routes/api-test.js [hunk 5]
```diff
@@ -18,7 +18,11 @@ describe('API', () => {
     },
     tasks: {
       addTask: sinon.stub(),
-      getTasks: sinon.stub()
+      getTasks: sinon.stub(),
+      changeTaskStatus: sinon.stub()
```
Comment (karimodm) [inline test/routes/api-test.js]: You don't use this

File: routes/api/tasks.js (hunk 2 above)
Comment (alongalky) [inline routes/api/tasks.js]: There are no tests for the database modules. I prefer to leave them as thin as possible

File: routes/api.js [hunk 6]
```diff
@@ -2,10 +2,11 @@ const express = require('express')
 const apiRouter = express.Router()
 const accountsRouter = express.Router({mergeParams: true})
 
-module.exports = ({machinesDatabase, tasksDatabase, cloud, tiers}) => {
+module.exports = ({accountsDatabase, machinesDatabase, tasksDatabase, cloud, tiers}) => {
```
Comment (alongalky) [inline routes/api.js]: Accepted. Refactoring

File: routes/api/tasks.js (hunk 1 above)
Comment (alongalky) [inline routes/api/tasks.js]: Refactored to avoid the problem

File: routes/api/tasks.js (hunk 3 above)
Comment (alongalky) [inline routes/api/tasks.js]: I agree, but scoping out of this PR. Added Trello task.

File: routes/api/tasks.js (hunk 4 above)
Comment (alongalky) [inline routes/api/tasks.js]: Thanks, but I refactored to avoid the issue

File: test/routes/api-test.js (hunk 5 above)
Comment (alongalky) [inline test/routes/api-test.js]: Nice catch.

Comment (alongalky) [inline test/routes/api-test.js]: Nice catch.

**Pull request merged**